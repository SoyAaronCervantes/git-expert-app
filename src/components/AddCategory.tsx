import React, {ChangeEvent, useState} from 'react';
import PropTypes from 'prop-types';
import {StateAction} from "../interfaces/StateAction";

const AddCategoryComponent = ({ setCategories }: StateAction<string[]> ) => {
  
  const [inputValue, setInputValue] = useState('');
  const handleInputChange = (e: ChangeEvent<HTMLInputElement> ) => {
    const value = e.target.value
    setInputValue( value );
  }
  
  const handleSubmit = (e: ChangeEvent<HTMLFormElement>) => {
    e.preventDefault();
    if ( inputValue.trim().length > 2 ) {
      setCategories( categories => [ inputValue, ...categories ] )
      setInputValue('');
    }
  };
  
  return (
    <form onSubmit={ handleSubmit }>
      <p>{ inputValue }</p>
      <input
        type="text"
        value={inputValue}
        onChange={ handleInputChange }
      />
    </form>
  );
};

AddCategoryComponent.propTypes = {
  setCategories: PropTypes.func.isRequired
};

export default AddCategoryComponent;
