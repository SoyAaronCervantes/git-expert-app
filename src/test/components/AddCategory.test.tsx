import React from 'react';
import { shallow, ShallowWrapper } from 'enzyme';
import AddCategory from '../../components/AddCategory';

describe('Test on AddCategory', () => {
  const setCategories = jest.fn()
  let wrapper: ShallowWrapper;

  beforeEach( () => {
    jest.clearAllMocks();
    wrapper = shallow(<AddCategory setCategories={ setCategories } /> );
  });

  test('should display correctly', () => {
    expect( wrapper ).toMatchSnapshot();
  });

  test('should change input', () => {

    const input = wrapper.find('input');
    const value = 'Hola mundo'

    input.simulate('change', { target: { value } });

    const paragraph = wrapper.find('p');
    const paragraphValue = paragraph.text().trim()

    expect( paragraphValue ).toBe( value );
  })

  test('should not post submit', () => {
    const form = wrapper.find('form');
    form.simulate('submit', { preventDefault: () => {} });

    expect( setCategories ).not.toHaveBeenCalled();
  });

  test('should call setCategory and clean input', () => {
    const value = "Hola mundo"

    const input = wrapper.find('input');
    input.simulate( 'change', { target: { value } } );

    const form = wrapper.find('form');
    form.simulate( 'submit', { preventDefault: () => {} })

    expect( setCategories ).toHaveBeenCalled();

    expect( input.prop('value') ).toBe('' );

  })

})