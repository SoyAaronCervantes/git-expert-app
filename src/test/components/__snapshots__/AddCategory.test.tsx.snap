// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test on AddCategory should display correctly 1`] = `
<form
  onSubmit={[Function]}
>
  <p />
  <input
    onChange={[Function]}
    type="text"
    value=""
  />
</form>
`;
